// dip-angular2/dist/device/instruments/osc/osc-channel.service.js
import { Injectable } from '@angular/core';
export var OscChannelService = (function () {
    function OscChannelService(oscChannelDescriptor) {
        this.effectiveBits = oscChannelDescriptor.effectiveBits;
        this.bufferSizeMax = oscChannelDescriptor.bufferSizeMax;
        this.bufferDataType = oscChannelDescriptor.bufferDataType;
        this.resolution = oscChannelDescriptor.resolution;
        this.sampleFreqMin = oscChannelDescriptor.sampleFreqMin;
        this.sampleFreqMax = oscChannelDescriptor.sampleFreqMax;
        this.adcVpp = oscChannelDescriptor.adcVpp;
        this.inputVoltageMax = oscChannelDescriptor.inputVoltageMax;
        this.inputVoltageMin = oscChannelDescriptor.inputVoltageMin;
        this.gains = oscChannelDescriptor.gains;
        this.delayMax = oscChannelDescriptor.delayMax;
        this.delayMin = oscChannelDescriptor.delayMin;
    }
    OscChannelService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    OscChannelService.ctorParameters = [
        null,
    ];
    return OscChannelService;
}());
