// dip-angular2/dist/device/instruments/awg/awg-channel.service.js
import { Injectable } from '@angular/core';
export var AwgChannelService = (function () {
    function AwgChannelService(awgChannelDescriptor) {
        this.signalTypes = awgChannelDescriptor.signalTypes;
        this.signalFreqMin = awgChannelDescriptor.signalFreqMin;
        this.signalFreqMax = awgChannelDescriptor.signalFreqMax;
        this.dataType = awgChannelDescriptor.dataType;
        this.bufferSizeMax = awgChannelDescriptor.bufferSizeMax;
        this.dacVpp = awgChannelDescriptor.dacVpp;
        this.dtMin = awgChannelDescriptor.dtMin;
        this.dtMax = awgChannelDescriptor.dtMax;
        this.vOffsetMin = awgChannelDescriptor.vOffsetMin;
        this.vOffsetMax = awgChannelDescriptor.vOffsetMax;
        this.vOutMin = awgChannelDescriptor.vOutMin;
        this.vOutMax = awgChannelDescriptor.vOutMax;
    }
    AwgChannelService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    AwgChannelService.ctorParameters = [
        null,
    ];
    return AwgChannelService;
}());
