// dip-angular2/dist/simulated-device/instruments/simulated-trigger.service.js
import { Injectable } from '@angular/core';
import { SimulatedDeviceHelperService } from '../simulated-device-helper.service';
export var SimulatedTriggerService = (function () {
    function SimulatedTriggerService(_simulatedDeviceService) {
        this.sources = [{
                "instrument": null,
                "channel": null,
                "type": null,
                "lowerThreshold": null,
                "upperThreshold": null
            }];
        this.targets = {};
        this.simulatedDeviceService = _simulatedDeviceService;
    }
    SimulatedTriggerService.prototype.setParameters = function (chan, source, targets) {
        this.sources[chan] = source;
        this.targets = targets;
        this.simulatedDeviceService.setTriggerTargets(targets);
        return {
            "command": "setParameters",
            "statusCode": 0,
            "wait": 0
        };
    };
    SimulatedTriggerService.prototype.single = function () {
        return {
            "command": "single",
            "statusCode": 0,
            "wait": -1,
            "acqCount": 27
        };
    };
    SimulatedTriggerService.prototype.stop = function () {
        return {
            "command": "stop",
            "statusCode": 0,
            "wait": -1
        };
    };
    SimulatedTriggerService.prototype.forceTrigger = function () {
        return {
            "command": "forceTrigger",
            "statusCode": 2684354589,
            "wait": -1
        };
    };
    SimulatedTriggerService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    SimulatedTriggerService.ctorParameters = [
        { type: SimulatedDeviceHelperService, },
    ];
    return SimulatedTriggerService;
}());
