// dip-angular2/dist/simulated-device/instruments/simulated-dc.service.js
import { Injectable } from '@angular/core';
import { SimulatedDeviceHelperService } from '../simulated-device-helper.service';
export var SimulatedDcService = (function () {
    function SimulatedDcService(_simulatedDeviceService) {
        this.voltages = [0, 0, 0, 0, 0, 0, 0, 0];
        this.simulatedDeviceService = _simulatedDeviceService;
    }
    SimulatedDcService.prototype.getVoltage = function (_chan) {
        return {
            command: 'getVoltage',
            voltage: this.voltages[_chan],
            statusCode: 0,
            wait: 100
        };
    };
    SimulatedDcService.prototype.setVoltage = function (_chan, _voltage) {
        this.voltages[_chan] = _voltage;
        return {
            command: 'setVoltage',
            statusCode: 0,
            wait: 0
        };
    };
    SimulatedDcService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    SimulatedDcService.ctorParameters = [
        { type: SimulatedDeviceHelperService, },
    ];
    return SimulatedDcService;
}());
