// dip-angular2/dist/data-types/waveform.service.js
import { Injectable } from '@angular/core';
export var WaveformService = (function () {
    function WaveformService(waveformDescriptor) {
        this.pointOfInterest = 0;
        this.triggerPosition = 0;
        this.seriesOffset = 0;
        this.triggerDelay = 0;
        {
            //Construct waveform from waveform descriptor object                   
            this.t0 = waveformDescriptor.t0;
            this.dt = waveformDescriptor.dt;
            this.y = waveformDescriptor.y;
            this.data = waveformDescriptor.data;
            this.pointOfInterest = waveformDescriptor.pointOfInterest;
            this.triggerPosition = waveformDescriptor.triggerPosition;
            this.seriesOffset = waveformDescriptor.seriesOffset;
            this.triggerDelay = waveformDescriptor.triggerDelay;
        }
    }
    WaveformService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    WaveformService.ctorParameters = [
        null,
    ];
    return WaveformService;
}());
